x-pack/plugins/session_view/public/components/ProcessTreeWindowing/hooks.ts

```diff
@@ -0,0 +1,183 @@
+    );
+
+    if (eventsPartition.exec.length) {
+      return eventsPartition.exec[eventsPartition.exec.length - 1];
+    }
+
+    if (eventsPartition.fork.length) {
+      return eventsPartition.fork[eventsPartition.fork.length - 1];
+    }
+
+    return this.events[this.events.length - 1] || ({} as ProcessEvent);
+  }
+
+  getOutput() {
+    return this.events.reduce((output, event) => {
+      if (event.event.action === EventAction.output) {
+        output += ''; // TODO: schema unknown
+      }
+
+      return output;
+    }, '');
+  }
+
+  isUserEntered() {
+    const event = this.getDetails();
+    const { interactive, pgid, parent } = event?.process || {};
+
+    return interactive && pgid !== parent.pgid;
+  }
+
+  getMaxAlertLevel() {
+    // TODO:
+    return null;
+  }
+}
+
+export const useProcessTree = ({
+  sessionEntityId,
+  forward,
+  backward,
+  searchQuery,
+}: UseProcessTreeDeps) => {
+  // initialize map, as well as a placeholder for session leader process
+  // we add a fake session leader event, sourced from wide event data.
+  // this is because we might not always have a session leader event
+  // especially if we are paging in reverse from deep within a large session
+  const fakeLeaderEvent = forward.find((event) => event.event.kind === EventKind.event);
+  const sessionLeaderProcess = new ProcessImpl(sessionEntityId);
+
+  if (fakeLeaderEvent) {
+    fakeLeaderEvent.process = { ...fakeLeaderEvent.process, ...fakeLeaderEvent.process.entry };
+    sessionLeaderProcess.events.push(fakeLeaderEvent);
+  }
+
+  const initializedProcessMap: ProcessMap = {
+    [sessionEntityId]: sessionLeaderProcess,
+  };
+
+  const [processMap, setProcessMap] = useState(initializedProcessMap);
+  const [forwardIndex, setForwardIndex] = useState(0);
+  const [backwardIndex, setBackwardIndex] = useState(0);
+  const [searchResults, setSearchResults] = useState<Process[]>([]);
+  const [orphans, setOrphans] = useState<Process[]>([]);
+
+  useEffect(() => {
+    let eventsProcessMap: ProcessMap = processMap;
+    if (backward) {
+      eventsProcessMap = processNewEvents(
+        eventsProcessMap,
+        backward.slice(0, backward.length - backwardIndex),
+        orphans,
+        sessionEntityId,
+        true
+      );
+      setBackwardIndex(backward.length);
+    }
+
+    eventsProcessMap = processNewEvents(
+      eventsProcessMap,
+      forward.slice(forwardIndex),
+      orphans,
+      sessionEntityId
+    );
+    setForwardIndex(forward.length);
+
+    setProcessMap({ ...eventsProcessMap });
+    setOrphans([...orphans]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [forward, backward]);
+
+  useEffect(() => {
+    setSearchResults(searchProcessTree(processMap, searchQuery));
+    autoExpandProcessTree(processMap);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [searchQuery]);
+
+  const flattenedSession = flattenProcessTree(processMap, sessionEntityId);
```

Comment:
We are trying to flatten the process tree here so that we can provide the length of the session view to `react-virtualized`  and for it to correctly locate which row to render

Reply:
The calculation of the length might get a bit "hacky" when we need to maintain the child processes expanded/collapsed and alerts detail expanded/collapsed.